pkg/client/logs.go

```diff
@@ -19,59 +19,119 @@ package client
 import (
 	"fmt"
 	"io"
-	"os"
-	"strings"
+	"sync"
 
-	"github.com/heptio/sonobuoy/pkg/config"
-	"github.com/sirupsen/logrus"
+	"github.com/pkg/errors"
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 )
 
-var (
-	podLogSeparator = strings.Repeat("-", 79)
-)
+// message represents a buffer of logs from a container in a pod in a namespace.
+type message struct {
+	// header acts as the header message for the log line and an id to know where the message came from.
+	header string
+	// buffer is the blob that we extracted from the contianer.
+	buffer []byte
+}
 
-// Logs gathers the logs for the containers in the sonobuoy namespace and prints them
+// logStreamer has all the pieces necessary for streaming logs from a container.
+type logStreamer struct {
+	ns, pod, container string
+	errc               chan error
+	logc               chan *message
+	logOpts            *v1.PodLogOptions
+	client             kubernetes.Interface
+}
 
-func (c *SonobuoyClient) GetLogs(cfg *LogConfig, client kubernetes.Interface) error {
-	// TODO(EKF): Stream to a writer instead of just stdout
-	if *cfg.Follow {
-		return streamLogs(client, cfg.Namespace, config.MasterPodName, &v1.PodLogOptions{Follow: true})
+// stream will open a connection to the pod's logs and push messages onto a channel.
+func (l *logStreamer) stream() {
+	req := l.client.CoreV1().Pods(l.ns).GetLogs(l.pod, l.logOpts)
+	readCloser, err := req.Stream()
+	if err != nil {
+		l.errc <- errors.Wrapf(err, "error streaming logs from container [%v]", l.container)
+		return
 	}
+	defer readCloser.Close()
+
+	header := fmt.Sprintf("## Namespace: %v ## Pod: %v ## Container: %v", l.ns, l.pod, l.container)
+
+	buf := make([]byte, 1024)
+	// Loop until EOF (streaming case won't get an EOF)
+	for {
+		n, err := readCloser.Read(buf)
+		if err != nil && err != io.EOF {
+			l.errc <- errors.Wrapf(err, "error reading logs from container [%v]", l.container)
+			return
+		}
+		l.logc <- &message{
+			header: header,
+			buffer: buf[:n],
+		}
+		if err == io.EOF {
+			return
+		}
+	}
+}
+
+// StreamLogs finds all pods in a given namespace and writes them to the configured io.Writer.
+func (c *SonobuoyClient) StreamLogs(cfg *LogConfig, client kubernetes.Interface) chan error {
```

Comment:
Could see an argument for returning a channel of messages as well as an error channel and force the consumer to do something with those channels.

Reply:
I would prefer a clean error with a tuple of channels.

Reply:
:+1: for `(chan Log, error)`

Reply:
Client should be all about communicating with Sonobuoy, presentation should be the CLI's responsibility

Reply:
signature will be (error, chan log, chan error)